Makes all objects in connection have a stable address

anarthal opened the pull request:
Moves read_buffer memory reservation to the connection's constructor
Makes read_buffer memory reservation size be a power of 2

mzimbres commented:
Hi Ruben, I am going to address those comments in the way you suggested. Just don't have much time now. I agree what you propose here is good (and common) practice and does not have any bad side effects. I support further development here, thanks.

anarthal commented:
Great. I will do some further fixes here and post it for review again.

I'm going to be out of office during the weekend too, so don't worry.

anarthal commented:
This is now ready @mzimbres. I've made all composed operations use `connection_impl`, and moved some of the functionality required by them there. I've also moved the reserve() call to the connection's constructor.

anarthal commented on include/boost/redis/connection.hpp:
```diff
@@ -382,16 +503,8 @@ class basic_connection {
-      set_receive_response(ignore);
-      writer_timer_.expires_at((std::chrono::steady_clock::time_point::max)());
-   }
+   : impl_(new detail::connection_impl<Executor>{std::move(ex), std::move(ctx), std::move(lgr)})
```
What would be the benefit? When using a unique_ptr as a variable, it saves you from spelling the type twice:

```cpp
std::unique_ptr<int> myint ( new int{42} );
auto myint = std::make_unique<int>(42);
```

But here it seems to add nothing:

```
impl_(new detail::connection_impl<Executor>{std::move(ex), std::move(ctx), std::move(lgr)})
impl_(std::make_unique<detail::connection_impl<Executor>>{std::move(ex), std::move(ctx), std::move(lgr)})
```

Note that `make_unique` doesn't improve efficiency as `make_shared` does.

mzimbres replied:
Correct, in this case I would only use it to avoid direct usages of the `new` operator.

anarthal commented on include/boost/redis/connection.hpp:
```diff
@@ -58,54 +58,157 @@ namespace boost::redis {
+
+   executor_type get_executor() noexcept { return writer_timer_.get_executor(); }
+
+   struct exec_op {
```
My long-term plan is to place all ops within connection_impl once they're small enough and using FSMs, but I didn't want to change too much.
